polkadot/runtime/parachains/src/assigner_on_demand/tests.rs

```diff
@@ -497,3 +536,127 @@ fn affinity_changes_work() {
+			// There is no affinity on the core prior to popping.
+			assert!(OnDemandAssigner::get_affinity_entries(*core_index).is_empty());
+
+			// There's always an order to be popped for each core.
+			let free_entries = OnDemandAssigner::get_free_entries();
+			let next_order = free_entries.peek();
+
+			// There is no affinity on the paraid prior to popping.
+			assert!(OnDemandAssigner::get_affinity_map(next_order.unwrap().para_id).is_none());
+
+			match OnDemandAssigner::pop_assignment_for_core(*core_index) {
+				Some(assignment) => {
+					// The popped assignment came from free entries.
+					assert_eq!(
+						start_free_entries - 1 - n,
+						OnDemandAssigner::get_free_entries().len()
+					);
+					// The popped assignment has the same para id as the next order.
+					assert_eq!(assignment.para_id(), next_order.unwrap().para_id);
+				},
+				None => panic!("Should not happen"),
+			}
+		});
+
+		// All entries have been removed from free_entries.
+		assert!(OnDemandAssigner::get_free_entries().is_empty());
+
+		// All chains have an affinity count of 1.
+		parachains.iter().for_each(|chain| {
+			assert_eq!(OnDemandAssigner::get_affinity_map(*chain).unwrap().count, 1);
+		});
+	});
+}
+
+#[test]
+#[should_panic]
+fn queue_index_ordering_is_unsound_over_max_size() {
+	// NOTE: Unsoundness proof. If the number goes sufficiently over the max_queue_max_size
+	// the overflow will cause an opposite comparison to what would be expected.
+	let mut max_num = u32::MAX - ON_DEMAND_MAX_QUEUE_MAX_SIZE;
+	// 0 < some large number.
+	assert_eq!(QueueIndex(0).cmp(&QueueIndex(max_num + 1)), Ordering::Less);
+}
+
+#[test]
+fn queue_index_ordering_works() {
+	// The largest number accepted in in the queue index.
+	let mut max_num = ON_DEMAND_MAX_QUEUE_MAX_SIZE;
+
+	// 0 == 0
+	assert_eq!(QueueIndex(0).cmp(&QueueIndex(0)), Ordering::Equal);
+	// 0 < 1
+	assert_eq!(QueueIndex(0).cmp(&QueueIndex(1)), Ordering::Less);
+	// 1 > 0
+	assert_eq!(QueueIndex(1).cmp(&QueueIndex(0)), Ordering::Greater);
+	// 0 < max_num
+	assert_eq!(QueueIndex(0).cmp(&QueueIndex(max_num)), Ordering::Less);
+	// 0 > max_num + 1
+	assert_eq!(QueueIndex(0).cmp(&QueueIndex(max_num + 1)), Ordering::Less);
+
+	// Ordering within the bounds of ON_DEMAND_MAX_QUEUE_MAX_SIZE works.
+	let mut v = vec![3, 6, 2, 1, 5, 4];
+	v.sort_by_key(|&num| QueueIndex(num));
+	assert_eq!(v, vec![1, 2, 3, 4, 5, 6]);
+
+	v = vec![max_num, 4, 5, 1, 6];
+	v.sort_by_key(|&num| QueueIndex(num));
+	assert_eq!(v, vec![1, 4, 5, 6, max_num]);
+
+	// Ordering with an element outside of the bounds of the max size also works.
+	v = vec![max_num + 2, 0, 6, 2, 1, 5, 4];
+	v.sort_by_key(|&num| QueueIndex(num));
+	assert_eq!(v, vec![0, 1, 2, 4, 5, 6, max_num + 2]);
+
+	// Numbers way above the max size are unsound and will fail to be sorted.
+	// Assume all numbers above the max size to be unsound.
+	v = vec![u32::MAX, 6, 2, 1, 5, 4];
+	v.sort_by_key(|&num| QueueIndex(num));
+	assert_eq!(v, vec![u32::MAX, 1, 2, 4, 5, 6]);
```

Comment:
This is actually "sound" and what we expect with the overflow. They are sorted and exactly how we want them to be sorted.

If the queue index is overflowing, then this sort is exactly what we want. Consider the following queue:

begin: u32::MAX - 1
end: 5

Entries in the queue would be: [u32::MAX-1, u32::MAX, 0, 1, 2, 3, 4] ... exactly what we want.